tree/treeplayer/inc/ROOT/TVecAllocator.hxx

```diff
@@ -0,0 +1,88 @@
+// Author: Enrico Guiraud, Enric Tejedor, Danilo Piparo CERN  01/2018
+
+/*************************************************************************
+ * Copyright (C) 1995-2018, Rene Brun and Fons Rademakers.               *
+ * All rights reserved.                                                  *
+ *                                                                       *
+ * For the licensing terms see $ROOTSYS/LICENSE.                         *
+ * For the list of contributors see $ROOTSYS/README/CREDITS.             *
+ *************************************************************************/
+
+#ifndef ROOT_TVECALLOCATOR
+#define ROOT_TVECALLOCATOR
+
+#include <iostream>
+#include <memory>
+
+namespace ROOT {
+namespace Detail {
+namespace VecOps {
+
+template <typename T>
+class TVecAllocator // : public std::allocator<T>
+{
+public:
+   using StdAlloc_t = std::allocator<T>;
+   using value_type = typename StdAlloc_t::value_type;
+   using pointer = typename StdAlloc_t::pointer;
+   using const_pointer = typename StdAlloc_t::const_pointer;
+   using reference = typename StdAlloc_t::reference;
+   using const_reference = typename StdAlloc_t::const_reference;
+   using size_type = typename StdAlloc_t::size_type;
+   using difference_type = typename StdAlloc_t::difference_type;
+
+private:
+   enum class EAllocType : char { kRegular, kFromExternalPointer, kNoneYet };
+   using StdAllocTraits_t = std::allocator_traits<StdAlloc_t>;
+   const_pointer fInitialAddress = nullptr;
+   size_type fInitialSize = 0;
+   EAllocType fAllocType = EAllocType::kRegular;
+   StdAlloc_t fStdAllocator;
+
+public:
+   TVecAllocator(const_pointer p, size_type n) : fInitialAddress(p), fInitialSize(n), fAllocType(EAllocType::kNoneYet){};
+   TVecAllocator() = default;
+   TVecAllocator(const TVecAllocator&) = default;
+
+   void construct(pointer p, const_reference val)
+   {
+      if (EAllocType::kFromExternalPointer == fAllocType)
+         return;
+      fStdAllocator.construct(p, val);
+   }
+
+   template <class U, class... Args>
+   void construct(U *p, Args &&... args)
+   {
+      if (EAllocType::kFromExternalPointer == fAllocType)
+         return;
+      fStdAllocator.construct(p, args...);
+   }
+
+   pointer allocate(std::size_t n)
+   {
+      if (n > std::size_t(-1) / sizeof(T))
```

Comment:
This will lead to warnings about mixing up negative values with unsigned types. Also, why throw here instead of leaving that to the real allocator? And why ignore `n` when reusing `fInitialAddress`? Will this not let people think they allocated more space when in fact they reused a pointer that has lower capacity?

Reply:
I don't think this prints any warnings, the comparison is between `std::size_t` and `std::size_t(-1)` is allowed afaik. That check comes from [cppreference](http://en.cppreference.com/w/cpp/concept/Allocator)'s demo allocator implementation by the way.